v23.1/performance-recipes.md

```diff
@@ -77,15 +84,109 @@ This section provides solutions for common performance issues in your applicatio
 
 #### Indicators that your application is experiencing transaction contention
 
-{% include {{page.version.version}}/performance/contention-indicators.md %}
+##### Waiting transaction
+
+These are indicators that a transaction is trying to access a row that has been ["locked"](architecture/transaction-layer.html#writing) by another, concurrent write transaction.
+
+- The **Active Executions** table on the **Transactions** page ([{{ site.data.products.db }} Console](../cockroachcloud/transactions-page.html) or [DB Console](ui-transactions-page.html#active-executions-table)) shows transactions with `Waiting` in the **Status** column. You can sort the table by **Time Spent Waiting**.
+- Querying the [`crdb_internal.cluster_locks`](crdb-internal.html#cluster_locks) table shows transactions where [`granted`](crdb-internal.html#cluster-locks-columns) is `false`.
+
+If a long-running transaction is waiting due to [contention](performance-best-practices-overview.html#transaction-contention): 
+
+1. [Identify the blocking transaction](#identify-conflicting-transactions). 
+1. Evaluate whether you can cancel the transaction. If so, [cancel it](#cancel-a-blocking-transaction) to unblock the waiting transaction.
+1. Optimize the transaction to [reduce further contention](#reduce-transaction-contention). In particular, break down larger transactions such as [bulk deletes](bulk-delete-data.html) into smaller ones to have transactions hold locks for a shorter duration, and use [historical reads](as-of-system-time.html) when possible to reduce conflicts with other writes.
+
+##### Transaction retry error
+
+These are indicators that a transaction has failed due to [contention](performance-best-practices-overview.html#transaction-contention).
+
+- A [transaction retry error](transaction-retry-error-reference.html) with `SQLSTATE: 40001`, the string [`restart transaction`](common-errors.html#restart-transaction), and an error code such as [`RETRY_WRITE_TOO_OLD`](transaction-retry-error-reference.html#retry_write_too_old) or [`RETRY_SERIALIZABLE`](transaction-retry-error-reference.html#retry_serializable), is emitted to the client.
+- An event with `TransactionRetryWithProtoRefreshError` is emitted to the CockroachDB [logs](logging-use-cases.html#example-slow-sql-query).
+
+{% include {{ page.version.version }}/performance/transaction-retry-error-actions.md %}
+
+##### Contention events
```

Comment:
Why do we have a separate section for contention events? Can we merge this with the "Waiting transaction" section? If so, this would be moved under the Transaction retry error section:

>The Transaction Restarts graph ([CockroachDB Cloud Console](https://deploy-preview-16823--cockroachdb-docs.netlify.app/docs/cockroachcloud/metrics-page#transaction-restarts) and [DB Console](https://deploy-preview-16823--cockroachdb-docs.netlify.app/docs/v23.1/ui-sql-dashboard#transaction-restarts) is showing spikes in retries over time.

Reply:
Done.